futurewei-cloud/arktos
# per-network service IP allocation - part 1: feature gate and preparation refactoring

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This is preparation for per-network service IP allocation:
1. introduces new feature gate PerNetworkServiceIPAlloc (off by default);
2. refactors api server code to call method instead of fetching field to get service IP alloc range object (which paves the way to per-network ip pool)

Per-network Service IP allocation is depicted in [service IP section, multi-tenant network design proposal](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md#service-ip).

**How to verify**:
in fact no change of behavior (unless the feature gate is explicitly turn on).
1. start arktos up by ```./hack/arktos-up.sh```
2. create a service (this works for now) with following yaml
```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-nginx-no-net
  labels:
    run: my-nginx
spec:
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: my-nginx
```
3. verify the named service created inside k8s service IP range
``` 
$ ./cluster/kubectl.sh get svc
NAME           TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes     ClusterIP   10.0.0.1     <none>        443/TCP   41m
my-nginx-no-net   ClusterIP   10.0.0.105   <none>        80/TCP    8m3s
```
If you want to further verify service that has explicit network, continue with
4. create network CRD by ```./cluster/kubectl.sh create -f ./pkg/controller/artifacts/crd-network.yaml```
5. create network test2 using yaml
```yaml
apiVersion: "arktos.futurewei.com/v1"
kind: Network
metadata:
  name: test2
spec:
  type: flat
  service:
    ipam: Arktos
```
6. create a service associated to test2 network, using yaml
```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-nginx-svc
  labels:
    run: my-nginx
    arktos.futurewei.com/network: test2
spec:
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: my-nginx
```
7.verify the service also gets IP address from k8s pool
```
$ ./cluster/kubectl.sh get svc my-nginx-svc
NAME           TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
my-nginx-svc   ClusterIP   10.0.0.185   <none>        80/TCP    26m
``` 

**Special notes for your reviewer**:
* This PR is intended to keep small and focused for ease of review. There will be another PR providing the method implementation and additional test code. 
* As this PR does not change behavior (when the feature gate if off by default), no new unit test cases is written; test code was changed only to reflect the newly refactored method in prod package.

**Does this PR introduce a user-facing change?**:
Partial: No by default; Yes potentially.

```release-note
If kube-apiserver start with --feature-gates=PerNetworkServiceIPAlloc=true, service IP address shall be allocated from network service IP pool (which is on the way)
```
